<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
<link rel="stylesheet" href="https://cdn.datatables.net/1.10.10/css/dataTables.bootstrap.min.css">


<h1><i class="fa fa-info-circle"></i>        What is TOXsIgN ?</h1>
<p>Briefly, the modern way of life results in humans as well as other living species to be exposed to numerous environmental components. The growing concerns about the adverse effects of these factors on a wide range of organisms,
  have led scientists to drastically increase the number of toxicological data. Moreover, considerable worries have recently been raised regarding the lack of reproducibility of biomedical research (<a href="https://www.ncbi.nlm.nih.gov/pubmed/22460859">1</a>), and more particularly in the field of toxicology (<a href="https://www.ncbi.nlm.nih.gov/pubmed/24747876">2</a>-<a href="https://www.ncbi.nlm.nih.gov/pubmed/25795653">3</a>–<a href="https://www.ncbi.nlm.nih.gov/pubmed/25149182">4</a>).
  Funding agencies, such as the National Institutes of Health (NIH), share this concern and discuss ways to enhance reproducibility in environmental sciences (5) by notably providing greater transparency of the data,
  including negative findings or contradictory data that should also be adopted more widely in peer-reviewed journals (<a href="https://www.ncbi.nlm.nih.gov/pubmed/24482835">5</a>).

While general repositories [such as the <a href="http://www.ncbi.nlm.nih.gov/geo">GEO</a> (<a href="https://www.ncbi.nlm.nih.gov/pubmed/23193258">6</a>) and the <a href="https://www.ebi.ac.uk/arrayexpress">ArrayExpress</a> (<a href="https://www.ncbi.nlm.nih.gov/pubmed/25361974">7</a>) databases] allow investigators to submit their raw data, other specialized databases
[such as the <a href="http://ctdbase.org">CTD</a> (<a href="https://www.ncbi.nlm.nih.gov/pubmed/25326323">8</a>), <a href="http://www.dixa-fp7.eu">diXa</a> (<a href="https://www.ncbi.nlm.nih.gov/pubmed/25505093">9</a>), <a href="http://toxdb.molgen.mpg.de">ToxDB</a> (<a href="https://www.ncbi.nlm.nih.gov/pubmed/27074805">10</a>) and <a href="http://www.drug2gene.com">Drug2Gene</a> (<a href="https://www.ncbi.nlm.nih.gov/pubmed/24618344">11</a>) resources] have paved the way for improving toxicological data storage, exchange and analysis (<a href="https://www.ncbi.nlm.nih.gov/pubmed/25538267">12</a>). Nevertheless, to the best of our knowledge, none of these resources allow scientists
to actively submit toxicological signatures, defined as the description of physiological, cellular, molecular or (transcript-/epigen-)omic effects on individuals or their descendants, after exposure to single or combined environmental factors,
including chemical (e.g. pesticides, plasticizers, drugs, endocrine disruptors), physical (e.g. radiations, temperature) or biological (e.g. pathogens, parasites) factors.

The TOXicological sIgNature (TOXsIgN) repository provides a flexible and open design that facilitates online submission, storage and retrieval of toxicological signatures deposited by the toxicology community.
One of the key features of TOXsIgN relies on its ability to archive heterogeneous data from: multiple species; observational and interventional studies; in vivo, ex vivo or in vitro experiments; physiological, molecular or “omic” studies;
transgenerational studies; and mixtures of environmental factors. Within TOXsIgN, Information is organized using eleven distinct ontologies in a four-layer architecture (project>study>assay>signature): a project is subdivided into studies
which address specific questions and describe treatments or experimental conditions (restricted to chemical factors in its current state), the latter being associated with toxicological signatures. TOXsIgN neither intends to archive raw data
such as GEO and ArrayExpress, nor to replace existing toxicological databases. Instead, it aims at complementing them by acting as a distribution hub. It is worth mentioning that password-protected access to prepublication data is also provided
for reviewers and authors.

In a near future, TOXsIgN will allow investigators to submit other types of environmental factors (physical and biological).

In addition to serving as a public archive, it intends to become a warehouse for toxicogenomics and predictive toxicology tools. TOXsIgN already includes a powerful search engine that supports complex fielded queries to retrieve data by accession number
of other various parameters such as species, tissues, developmental stage, chemicals, type of experiment, observed toxicological effects. The modular design of TOXsIgN will facilitate the implementation of other advanced tools
(currently in development at the IRSET) leaning on the deposited toxicogenomic signatures that will help investigators analyze, predict and prioritize the toxicological effects of environmental factors relevant to their specific interests.


<ul>
  <li>1. Must try harder. (2012) Nature, 483, 509.</li>
  <li>2. Miller,G.W. (2014) Improving reproducibility in toxicology. Toxicol. Sci., 139, 1–3.</li>
  <li>3. George,B.J., Sobus,J.R., Phelps,L.P., Rashleigh,B., Simmons,J.E., Hines,R.N. and Community of Practice for Statistics Guidance Documents Working Groups (2015) Raising the bar for reproducible science at the U.S. Environmental Protection Agency Office of Research and Development. Toxicol. Sci., 145, 16–22.</li>
  <li>4. Poland,C.A., Miller,M.R., Duffin,R. and Cassee,F. (2014) The elephant in the room: reproducibility in toxicology. Part. Fibre Toxicol., 11, 42.</li>
  <li>5. Collins,F.S. and Tabak,L.A. (2014) Policy: NIH plans to enhance reproducibility. Nature, 505, 612–3.</li>
  <li>6. Barrett,T., Wilhite,S.E., Ledoux,P., Evangelista,C., Kim,I.F., Tomashevsky,M., Marshall,K.A., Phillippy,K.H., Sherman,P.M., Holko,M., et al. (2013) NCBI GEO: archive for functional genomics data sets--update. Nucleic Acids Res., 41, D991–5.</li>
  <li>7. Kolesnikov,N., Hastings,E., Keays,M., Melnichuk,O., Tang,Y.A., Williams,E., Dylag,M., Kurbatova,N., Brandizi,M., Burdett,T., et al. (2015) ArrayExpress update--simplifying data submissions. Nucleic Acids Res., 43, D1113–6.</li>
  <li>8. Davis,A.P., Grondin,C.J., Lennon-Hopkins,K., Saraceni-Richards,C., Sciaky,D., King,B.L., Wiegers,T.C. and Mattingly,C.J. (2015) The Comparative Toxicogenomics Database’s 10th year anniversary: update 2015. Nucleic Acids Res., 43, D914–20.</li>
  <li>9. Hendrickx,D.M., Aerts,H.J.W.L., Caiment,F., Clark,D., Ebbels,T.M.D., Evelo,C.T., Gmuender,H., Hebels,D.G.A.J., Herwig,R., Hescheler,J., et al. (2015) diXa: a data infrastructure for chemical safety assessment. Bioinformatics, 31, 1505–7.</li>
  <li>10. Hardt,C., Beber,M.E., Rasche,A., Kamburov,A., Hebels,D.G., Kleinjans,J.C. and Herwig,R. (2016) ToxDB: pathway-level interpretation of drug-treatment data. Database (Oxford)., 2016.</li>
  <li>11. Roider,H.G., Pavlova,N., Kirov,I., Slavov,S., Slavov,T., Uzunov,Z. and Weiss,B. (2014) Drug2Gene: an exhaustive resource to explore effectively the drug-target relation network. BMC Bioinformatics, 15, 68.</li>
  <li>12. Miller,G.W. (2015) Data sharing in toxicology: beyond show and tell. Toxicol. Sci., 143, 3–5.</li>
</ul>



<h1><i class="fa fa-book"></i>      How to cite TOXsIgN ?</h1>
<p>If TOXsIgN has been useful to your work, please consider citing our latest publication: Darde T.A., Gaudriault P., Beranger R., Lancien C., Caillarec-Joly A., Sallou O., Bonvallot N., Chevrier C., Mazaud-Guittot S., Jégou B., Collin O., Becker E., Rolland A.D., Chalmel F. <a href="https://www.ncbi.nlm.nih.gov/pubmed/29385404" target="_blank">TOXsIgN: a cross-species repository for toxicogenomic signatures, Bioinformatics</a>.</p>

<h1><i class="fa fa-users"></i>      Who has supported TOXsIgN?</h1>
<p>The TOXsIgN repository has been funded via grants from the <a href="https://www.anses.fr/en">French Agency for Food, Environmental and Occupational Health & Safety (ANSES)</a> [No. EST-13-081 to F.C.] and the Medical Research Foundation (FRM) [No. DBI20131228558 to F.C.], the European Union [FEDER to F.C.].
  It is supported, built and maintained by the Research Institute for Environmental and Occupational Health (IRSET), the French School of Public Health (EHESP) and the GenOuest BioInformatics core facility located in Rennes (Brittany, France). The TOXsIgN repository is and will remain under constant development to offer more tools and enhance user experience.</p>

<h1><i class="fa fa-envelope-o"></i>      Contact Us</h1>
<div class="row">
  <div class="col-md-6">
    Frédéric Chalmel<br>
    INSERM U1085-Irset<br>
    University of Rennes 1<br>
    Campus de Beaulieu, Bâtiment 13<br>
    263 avenue du Général Leclerc<br>
    35042 Rennes<br>
    France<br>
    Phone: +33 (0)2 23 23 58 02<br>
    Fax: +33 (0)2 23 23 50 55<br>
    Mail: frederic.chalmel@inserm.fr<br>
  </div>

  <div class="col-md-6">
    GenOuest BioInformatics Platform<br>
    Campus de Beaulieu, IRISA-INRIA<br>
    263 avenue du Général Leclerc<br>
    35042 Rennes<br>
    France<br>
    Phone: +33 (0)2 99 84 72 78<br>
    Fax: +33 (0)2 99 84 71 71<br>
    Mail: support@genouest.org<br>
  </div>
</div>

<h1><i class="fa fa-wrench"></i>      Future developments</h1>
<p>We are currently integrating additional toxicogenomic signatures from other major toxicogenomics projects, such as <a href="http://www.lincsproject.org/">LINCS</a>, <a href="https://portals.broadinstitute.org/cmap/">CMAP</a>, <a href="http://www.dixa-fp7.eu/">diXa</a>
  and <a href="https://tools.niehs.nih.gov/cebs3/ui/">CEBS</a> (Lamb et al., 2006; Lea et al., 2017; Hendrickx et al., 2015). Likewise, we plan to make TOXsIgN compatible with other kind of environmental agents,
   as physical (e.g. radiations, temperature) and biological (e.g. pathogens, parasites) factors. Altogether, we expect that this new resource can contribute significantly to risk assessment.
In addition to serving as a public repository, TOXsIgN is also intended to be a warehouse for toxicogenomic and predictive toxicology tools. Its modular design facilitates the implementation of additional bioinformatics modules relying on
the deposited toxicogenomic signatures that will help investigators analyze and predict adverse effects of environmental factors relevant to their specific interests.
We are currently developing prediction and prioritization systems for chemical toxicity. Another potential aspect is the addition of social features in TOXsIgN to enable several investigators to work on the same data.
Finally, we plan to incorporate the ISA framework (<a href="http://isa-tools.org/">http://isa-tools.org/</a>), which provides rich descriptions of experimental metadata, to manage the TOXsIgN database and improve the FAIRness of available data.
Together these efforts will improve TOXsIgN’s utility, making it a frontline resource in toxicogenomics.

</p>
<table class="table table-bordered table-hover table-striped">
          <tr>
            <th>Type</th>
            <th>Description</th>
            <th>Progression</th>
          </tr>
          <tr>
              <td>Tool</td>
              <td>Chemical Prioritization Systeme (ChemPSy)</td>
              <td><div class="progress-bar progress-bar-info" role="progressbar" aria-valuenow="70" aria-valuemin="0" aria-valuemax="100" style="width:70%"> 70% Complete </div></td>
          </tr>
          <tr>
              <td>Tool</td>
              <td>Proximity automatic calcul between toxicogenomic signatures </td>
              <td><div class="progress-bar progress-bar-warning" role="progressbar" aria-valuenow="100" aria-valuemin="0" aria-valuemax="100" style="width:100%"> 100% Complete </div></td>
          </tr>
          <tr>
              <td>Display feature</td>
              <td>Enhance search results visualization</td>
              <td><div class="progress-bar progress-bar-warning" role="progressbar" aria-valuenow="20" aria-valuemin="0" aria-valuemax="100" style="width:100%"> 100%  </div></td>
          </tr>
          <tr>
              <td>Display feature</td>
              <td>User friendly project/study/signature interface</td>
              <td><div class="progress-bar progress-bar-info" role="progressbar" aria-valuenow="50" aria-valuemin="0" aria-valuemax="100" style="width:100%"> 100% Complete </div></td>
          </tr>
</table>
