<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">

<h2><i class="fa fa-search" aria-hidden="true"></i>    Search</h2>
<div class="panel-group" id="accordion">
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion" href="#searchengine" target="_self">
        Search engine</a><a class="pull-right" href="#/query">use</a>
      </h4>
    </div>
    <div id="searchengine" class="panel-collapse collapse in">
      <div class="panel-body">The search engine supports complex fielded queries to retrieve signatures and associated data based on various information such as experimental parameters, observed toxicological effects and/or up/down-regulated gene lists.
        This module is especially powerfull for identifying comparable experiments (i.e. performed on similar models and/or with close environmental factors) and/or similar outcomes (i.e. close physiological, molecular and/or omics effects).</div>
    </div>
  </div>
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion" href="#ontologies" target="_self">
        Browse ontologies</a><a class="pull-right" href="#/ontologies">use</a>
      </h4>
    </div>
    <div id="ontologies" class="panel-collapse collapse">
      <div class="panel-body">TOXsIgN also includes a tool to browse the ontologies used in the resource to describe data.
        This feature is especially relevant for filling the excel template required when submitting a novel signature. This tool use the NCBI OntologyTree widget from BioPortal. It is configured to display all the ontologies used in TOXsIgN.</div>
  </div>
</div>

<hr>


<h2><i class="fa fa-exchange" aria-hidden="true"></i>    Convert</h2>
<div class="panel-group" id="accordion2">
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion2" href="#convert" target="_self">
        Conversion tool</a><a class="pull-right" href="#/convert">use</a>
      </h4>
    </div>
    <div id="convert" class="panel-collapse collapse">
      <div class="panel-body">The conversion tool allows users to convert a list of EntrezGene Ids or HomoloGene Ids to HomoloGene Ids or EntrezGene Ids.</div>
    </div>
  </div>
</div>



<hr>


<h2><i class="fa fa-compress" aria-hidden="true"></i>    Compare</h2>
<div class="panel-group" id="accordion3">
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion3" href="#signature" target="_self">
        Signature Comparison (Venn diagram)</a><a class="pull-right" href="#/compare">use</a>
      </h4>
    </div>
    <div id="signature" class="panel-collapse collapse">
      <div class="panel-body">The comparison module allows users to compare several toxicogenomic signatures It is based on the tool developed by P. Bardou et al that can display intersection between up to 6 signatures in a Venn diagram.</div>
    </div>
  </div>
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion3" href="#dist" target="_self">
        Signature enrichment analysis</a><a class="pull-right" href="#/dist">use</a>
      </h4>
    </div>
    <div id="dist" class="panel-collapse collapse">
      <div class="panel-body">The signature enrichment analysis module allows users to compare a selected toxicogenomic signatures to the overall set of signatures deposited by the community using a distance analysis approach (based on the Euclidean distance and the correlation coefficient).</div>
    </div>
  </div>
</div>

<hr>


<h2><i class="fa fa-eye" aria-hidden="true"></i>    Analyze</h2>
<div class="panel-group" id="accordion4">
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion4" href="#function" target="_self">
        Functional enrichment analysis</a><a class="pull-right" href="#/enrich">use</a>
      </h4>
    </div>
    <div id="function" class="panel-collapse collapse">
      <div class="panel-body">The functional enrichment analysis allows users to identify biological processes, molecular functions, cellular components and phenotypes significantly associated to a selected signature.</div>
    </div>
  </div>
  <div class="panel panel-default">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion4" href="#dist2" target="_self">
        Signature enrichment analysis</a><a class="pull-right" href="#/dist">use</a>
      </h4>
    </div>
    <div id="dist2" class="panel-collapse collapse">
      <div class="panel-body">The signature enrichment analysis module allows users to compare a selected toxicogenomic signatures to the overall set of signatures deposited by the community using a distance analysis approach (based on the Euclidean distance and the correlation coefficient).</div>
    </div>
  </div>
</div>

<hr>


<h2><i class="fa fa-external-link" aria-hidden="true"></i>    Extract</h2>
<div class="panel-group" id="accordion5">
  <div class="panel panel-danger">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion5" href="#extract" target="_self">
        Extract toxicogenomic signature (under construction)</a>
      </h4>
    </div>
    <div id="extract" class="panel-collapse collapse">
      <div class="panel-body">This module allows users to automatically extract the toxicogenomic signature from a transcriptomic dataset (Affymetrix GeneChip). This tool is under construction and will be available soon.</div>
    </div>
  </div>
</div>


<hr>


<h2><img src="images/crystalball.png" style="width: 1em;">    Predict</h2>
<div class="panel-group" id="accordion6">
  <div class="panel panel-danger">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion6" href="#predict" target="_self">
        ChemPSy - Prediction tool (under construction)</a>
      </h4>
    </div>
    <div id="predict" class="panel-collapse collapse">
      <div class="panel-body">This module will allow users to predict the adverse effects of a given chemical compound based on the approaches developed in the ChemPSy project.
The aim of ChemPSy (Chemical Prioritization System) is to provide an innovative tool based on several bioinformatics and biostatistics methodologies to analyze and integrate massive toxicogenomic datasets. Specific objectives include: (1) classification of chemicals based on transcriptional signatures, i.e. the set of genes whose expression is known to be positively or negatively altered after an exposure to these compounds; (2) the association of classes with human pathologies or deleterious phenotypes, i.e. classes containing toxicants with well-known effects; (3) the prediction of novel reprotoxicants and/or endocrine disruptors based on transcriptional signature similarities with known chemicals affecting testis development and function. This tool is under construction and will be available soon.
</div>
    </div>
  </div>
  <div class="panel panel-danger">
    <div class="panel-heading">
      <h4 class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion6" href="#prio" target="_self">
        ChemPSy - Prioritization tool (under construction)</a>
      </h4>
    </div>
    <div id="prio" class="panel-collapse collapse">
      <div class="panel-body">This module will allow users to prioritize several selected chemical compounds based on the approaches developed in the ChemPSy project.
The aim of ChemPSy (Chemical Prioritization System) is to provide an innovative tool based on several bioinformatics and biostatistics methodologies to analyze and integrate massive toxicogenomic datasets. Specific objectives include: (1) classification of chemicals based on transcriptional signatures, i.e. the set of genes whose expression is known to be positively or negatively altered after an exposure to these compounds; (2) the association of classes with human pathologies or deleterious phenotypes, i.e. classes containing toxicants with well-known effects; (3) the prediction of novel reprotoxicants and/or endocrine disruptors based on transcriptional signature similarities with known chemicals affecting testis development and function. This tool is under construction and will be available soon. </div>
    </div>
  </div>
</div>


<!--<li ng-if="is_logged" class="btn-group" dropdown is-open="status.isopen">
                <a type="button" dropdown-toggle ng-disabled="disabled">
                  Tools<span class="caret"></span>
                </a>
                <ul class="dropdown-menu" role="menu">

                  <li class="dropdown-header">Predict</li>
                  <li><a href="#/predict"> ChemPSy - Prediction tool</a></li>
                  <li><a href="#/prio"> ChemPSy - Prioritization tool</a></li>
                </ul>
              </li>-->
